### DISCUSSION TITLE
提供PAC代理功能

### DISCUSSION THREAD

**Discussion by giveup** [SeaHOH/GotoX#13] [点子]
初衷应该是设置为系统全局代理,然后再根据IP地址决定是否走代理.
这样可能会有更多本来能直连的IP也走代理了.感觉不如PAC好.

**Comment by SeaHOH** [comment]
就是觉得 pac 不好用才做的这个，客户端没有负担，使用很方便。

这个匹配的不是 IP 而是主机名，你可以看看这个规则说明 [ActionFilter.ini](https://github.com/SeaHOH/GotoX/blob/master/config/ActionFilter.ini)。

设定了规则的链接按规则走，如果没有匹配，默认配置是这样：
**HTTP** 链接由代理直接请求资源后再将数据返回
**HTTPS** 链接由代理直接转发链接
这两种都可以称为直连，完全不走代理。

有特需求的话，还可以自定义默认配置为其它方式。
可以轻易的实现黑白名单，也可进行各种复杂匹配。繁简由心。

如果你要使用 pac 也可以，有一个专门的 GAE 端口，你可以自己维护一个 pac 来使用它。

**Comment by giveup** [comment]
@SeaHOH 抱歉,没理解好.
测试了下,所以链接都由代理处理的话,包括那些直连的,容易出现timeout,所以我想不通过代理直连可能会好一点.

**Comment by SeaHOH** [comment]
因为超时参数故意调的比较低，你可以在主配置中调整。
不过反正有重试机制，我觉得影响不大。

**Comment by SeaHOH** [comment]
@giveup 
我重新调整了代码，将各类链接的超时参数彻底分离，这下更容易自定义设置了。
你可以下载主分支来试一下，可分别设置 `GAE`、`普通（直连）链接` 和 `转发（其它后端代理）链接` 的超时参数。

**Comment by giveup** [comment]
@SeaHOH 出现 Broken pipe 的话应该调整什么参数合适？

13:57:06 I 17.252.201.246 "FWD CONNECT api-aka.smoot.apple.com:443 HTTP/1.1" - -
13:57:18 W 转发 None 失败：BrokenPipeError(32, 'Broken pipe')

21:20:51 W 113.171.253.26 do_GAE "POST https://gsp64-ssl.ls.apple.com/a/v2/use" 返回：BrokenPipeError(32, 'Broken pipe')，重试
21:20:52 E 113.171.240.110 do_GAE "POST https://gsp64-ssl.ls.apple.com/a/v2/use" 失败：BrokenPipeError(32, 'Broken pipe')`

21:25:14 I 192.30.253.116 "FWD CONNECT api.github.com:443 HTTP/1.1" - -
21:25:20 W 转发 None 失败：BrokenPipeError(32, 'Broken pipe')
21:25:20 W 转发 None 失败：BrokenPipeError(32, 'Broken pipe')
21:25:29 I 17.248.156.108 "FWD CONNECT p44-keyvalueservice.icloud.com:443 HTTP/1.1" - -
21:25:31 I 17.248.156.108 "FWD CONNECT p44-keyvalueservice.icloud.com:443 HTTP/1.1" - -
21:25:33 W 转发 None 失败：BrokenPipeError(32, 'Broken pipe')
21:25:33 I 192.30.253.113 "FWD CONNECT github.com:443 HTTP/1.1" - -
21:26:16 W 转发 None 失败：BrokenPipeError(32, 'Broken pipe')

**Comment by ghost** [comment]
可以参照 shadowrocket 的规则修改智能代理么
https://github.com/h2y/Shadowrocket-ADBlock-Rules/blob/master/sr_top500_whitelist_ad.conf
https://github.com/Hsiny/Shadowrocket-Config/blob/master/simplifyrule.txt
智能代理是PAC 2.0 , 希望多多支持

**Comment by SeaHOH** [comment]
@blob8964 今天第二个交流~
现在这种时期居然还有人在用，我太感动了~

不过呢，现在可能要打击到你了。
因为我不打算加入太多默认规则，尤其是去广告，你可以参照说明自己添加。

**Comment by ghost** [comment]
@seahoh
因为surge的智能代理功能,gotox基本都支持,所以选择goto
不过 actionfilter.ini 没给出ip段格式
1.9.22.0/24-1.9.33.0/24
1.9.22.0-255
这种i[p段格式支持么

**Comment by SeaHOH** [comment]
可以使用正式表达式匹配：
`@^1\.9\.2[23]\.\d{1-3}$`
第二条如果不严格匹配可使用域名规则：
`1.9.22.`

但是，规则匹配是针对网址的，如果没有域名，IP 就是主机名，想匹配域名解析后的 IP 是不行的。
IP 属于另一层面，我只做了直连判断（国内和保留地址），没有相应规则，感觉没必要。

**Comment by wz7465** [comment]
说说一些设计上的问题
8087是自动代理，8086是gae的pac模式，我有gotox的智能路由了为何还要pac模式？功能定位重复，我想使用gae的全代理功能却没法实行，访问github还是走的国内直连，其实通过gae访问更快，另外如果发帖的话，还暴露自己的ip
自动代理重复做功，程序关掉后，所有暂时保存的路由全部归零，再次打开时还得重新建立。应该保存一份动态的用户路由列表，下次程序打开时，优先读取该路由

**Comment by SeaHOH** [comment]
https://github.com/SeaHOH/GotoX/issues/13#issuecomment-289030620
请看上面，有自动代理端口的规则说明。

其实我是不好意思叫智能代理的，因为真的不智能，经常要手动修改规则。

**Comment by wz7465** [comment]
你说的是这里修改吗？
[50-gae] #通过 GAE 服务器代理
#通过 GAE 才能正常使用的 google 服务
#counts.google.com
.github.com （添加）

**Comment by SeaHOH** [comment]
嗯嗯，还有其它几个域名，想一起走 GAE 也可以。

不过我一般放最后 `90` 那个，`50` 这些数字就是指匹配顺序。

**Comment by wz7465** [comment]
事实上我添加了，它仍然没成功走gae，还是通过自动代理走

**Comment by SeaHOH** [comment]
没有输出像这样的？注意 GAE 字样。
>14:40:23 I L4:55723->151.101.36.133 "GAE GET https://assets-cdn.github.com/images/icons/emoji/unicode/270b.png HTTP/1.1" 200 4256

**Comment by wz7465** [comment]
193.87.134.38 39是sni

![default](https://user-images.githubusercontent.com/7801728/38916106-ae8bbaf2-4318-11e8-9233-7c7ce46535db.png)

**Comment by wz7465** [comment]
就放了两个ip，连接GitHub.com依然有直连

**Comment by SeaHOH** [comment]
嗯，规则可以这样写：
```
github.com$
```
或者
```
github.com
.github.com
```

**Comment by wz7465** [comment]
感觉好点，不过一个网站涉及多个域名，有很多直连，应该还存在本地ip暴露的风险，如果流量全走一个gae ip，应该好点

**Comment by SeaHOH** [comment]
有单独的 GAE 端口，但是你不可能全走它啊，国内那么多网站。
gfwlist 是规则，这个也是，只是需要自己动手。

**Comment by wz7465** [comment]
你的gae端口是pac模式的，不如搞成全局模式，这样可以快速手动切换一下，我要速度时就用自动代理，想匿名时就用gae 全局模式

**Comment by SeaHOH** [comment]
8086

**Comment by SeaHOH** [comment]
注意，其实没有匿名，首先就暴露了 AppID。

**Comment by wz7465** [comment]
8086端口我试过了，还是国内国外自动分流，改成国内国外流量全走gae代理吧
![default](https://user-images.githubusercontent.com/7801728/38918030-156586d0-431f-11e8-87ec-7c0c38bd5892.png)

**Comment by SeaHOH** [comment]
你的是什么浏览器？怎么配置代理的？

**Comment by wz7465** [comment]
火狐，插件foxyproxy

**Comment by SeaHOH** [comment]
检查 foxyproxy 规则

**Comment by SeaHOH** [comment]
记住，真的没匿名，再提醒一次。

**Comment by wz7465** [comment]
误会了，8086端口是全局模式，gotox设置按钮没起作用，我手动添加一个8086端口
另，需要的匿名也是有限匿名，比如我访问github，只要服务器上显示的访问地址是谷歌IP就行了，对谷歌来说肯定不匿名

**Comment by SeaHOH** [comment]
谷歌不光知道你 IP，还知道你的 urls、cookies。

**Comment by wz7465** [comment]
如今中美蜜月期，是该悠着点，不过少涉及正治，应该没事啦，多谢
